### PR TITLE
refactor(driver): move reduce logic to BaseDriver

### PR DESCRIPTION
**Pull request by hanxiao**
WIP, otherwise `ReduceDriver` is orthogonal to all other Drivers.
also fixing `wrap_func` in `register_class`

### CODE REVIEW

File: jina/drivers/reduce.py [hunk 1]
```diff
@@ -113,42 +50,31 @@ def _apply_all(
...
-            doc_pointers[context_doc.id] = context_doc.evaluations
+        if context_doc.id not in self.doc_pointers:
+            self.doc_pointers[context_doc.id] = context_doc.evaluations
         else:
-            doc_pointers[context_doc.id].extend(context_doc.evaluations)
+            self.doc_pointers[context_doc.id].extend(context_doc.evaluations)
 
 
-class ConcatEmbedDriver(ReduceDriver):
+class ConcatEmbedDriver(ReduceAllDriver):
```
Comment (JoanFM) [inline jina/drivers/reduce.py:61]: I think `ConcatEmbedDriver` should either disappear and changed by a `ConcatEncoder` which would be a `MultiModalEncoder` or at least separate it from the `ReduceAllDriver` (as u say the logic is orthogonal), one may want to concat the `embedding` of 2 chunks?

Comment (JoanFM) [reply jina/drivers/reduce.py:61]: If moved to a `MultiModalEncoder` then it would need to be combined with a `uses_before` = reduce

Comment (hanxiao) [reply jina/drivers/reduce.py:61]: make sense, but right now in the core I can only find `ConcatEncoder` in test: `tests/integration/multimodal/yaml/embedding-multimodal-concatenate.py`? what's the plan here?

**Comment by JoanFM** [comment]
Is it needed? Since we added the option to use `uses_before` it is quite flexible and easy to use separatedly.  I think is nice to have only one class for testing, or add it as another `mixin`

File: jina/drivers/reduce.py (hunk 1 above)
Comment (JoanFM) [reply jina/drivers/reduce.py:61]: We added this just for testing, we could add as a real executor in the `hub`, it can be interesting.

Comment (JoanFM) [reply jina/drivers/reduce.py:61]: Actually we were desigining `MultiModalDriver` the same way `ConcatEmbedDriver` is done, but then we realised we needed to separate these concerns.

**Comment by hanxiao** [comment]
> Is it needed? Since we added the option to use `uses_before` it is quite flexible and easy to use separatedly. I think is nice to have only one class for testing, or add it as another `mixin` maybe?

the idea here is to allow all drivers to access `partial_reqs` and apply their own logic to all `partial_reqs`, providing a uniform interface to all drivers.

If you look at the `__apply_all__` and `traverse_` logic before, there are some redundancy and inconsistency to other drivers.

File: jina/drivers/reduce.py (hunk 1 above)
Comment (hanxiao) [reply jina/drivers/reduce.py:61]: ConcatEmbed is one level, group by `doc.id` only, `MultiModalDriver` group by `doc.id` and `doc.modality` if I understand correctly

Comment (JoanFM) [reply jina/drivers/reduce.py:61]: yes, it has some slight differences right

File: jina/drivers/__init__.py [hunk 2]
```diff
@@ -134,15 +137,27 @@ class BaseDriver(metaclass=DriverType):
 
     store_args_kwargs = False  #: set this to ``True`` to save ``args`` (in a list) and ``kwargs`` (in a map) in YAML config
 
-    def __init__(self, priority: int = 0, *args, **kwargs):
+    def __init__(self, priority: int = 0, expect_parts: Optional[int] = 1, *args, **kwargs):
```
Comment (JoanFM) [inline jina/drivers/__init__.py:140]: Remember to add docstring for num parts

Comment (JoanFM) [reply jina/drivers/__init__.py:140]: Now a user needs to know how many parts it expects? It seems worse than before when user did not have to worry

**Review by JoanFM — COMMENTED** [review]
I think maybe my messages come from my lack of clarity on the difference between ReduceDriver and ReduceAllDriver, I will check in more detail

File: jina/drivers/__init__.py (hunk 2 above)
Comment (hanxiao) [reply jina/drivers/__init__.py:140]: no need, but this is for preparing #1215

File: jina/drivers/reduce.py (hunk 1 above)
Comment (hanxiao) [reply jina/drivers/reduce.py:61]: no, they are irrelevant. `ConcatEmbedDriver` is a `ReduceDriver` which works on multiple requests. Complete different task than `MultiModalDriver`

Comment (JoanFM) [reply jina/drivers/reduce.py:61]: But it could be able to concatenate chunks coming from the same segmenter without the need of reducing.

Comment (JoanFM) [reply jina/drivers/reduce.py:61]: Anyway no need to change now

**Pull request merged**